miden-lib/asm/sat/internal/faucet.masm

```diff
@@ -0,0 +1,281 @@
+    # => [ASSET]
+
+    # add the asset to the input vault for asset preservation checks
+    dupw exec.layout::get_input_vault_root_ptr movdn.4 exec.asset_vault::add_fungible_asset dropw
+    # => [ASSET]
+end
+
+#! Burns a fungible asset associated with the fungible faucet the transaction is being executed
+#! against.
+#!
+#! Panics:
+#! - If the transaction is not being executed against a fungible faucet.
+#! - If the fungible asset being burned is not associated with the faucet the transaction is
+#!   being executed against.
+#! - If the asset is not well formed.
+#! - If the amount being burned is greater than the total input to the transaction.
+#!
+#! Stack: [ASSET]
+#! Outputs: [ASSET]
+#!
+#! - ASSET is the asset that was burned.
+proc.burn_fungible_asset
+    # assert that we are executing a transaction against a fungible faucet (access checks)
+    exec.account::get_id dup exec.account::is_fungible_faucet assert
+    # => [fungible_faucet_id, ASSET]
+
+    # assert that the asset is associated with the faucet the transaction is being executed against
+    dup.1 assert_eq
+    # => [ASSET]
+
+    # assert the asset is well formed
+    dup.1 not assert
+    dup.2 not assert 
+    # => [ASSET]
+
+    # fetch TOTAL_ISSUANCE such that we can compute the new total issuance
+    push.FAUCET_DATA_STORAGE_SLOT exec.account::get_item
+    # => [TOTAL_ISSUANCE, ASSET]
+
+    # compute new total issuance
+    dup.7 sub push.FAUCET_DATA_STORAGE_SLOT exec.account::set_item dropw
+    # => [ASSET]
+
+    # remove the asset from the input vault
+    dupw exec.layout::get_input_vault_root_ptr movdn.4 exec.asset_vault::remove_fungible_asset dropw
+    # => [ASSET]
+end
+
+# NON-FUNGIBLE ASSETS
+# ==================================================================================================
+
+#! Mints a non-fungible asset associated with the non-fungible faucet the transaction is being
+#! executed against.
+#!
+#! Panics:
+#! - If the transaction is not being executed against a non-fungible faucet.
+#! - If the non-fungible asset being minted is not associated with the faucet the transaction is
+#!   being executed against.
+#! - If the non-fungible asset being minted already exists.
+#!
+#! Stack: [ASSET]
+#! Outputs: [ASSET]
+#!
+#! - ASSET is the asset that was minted.
+proc.mint_non_fungible_asset
+    # assert that we are executing a transaction against the non-fungible faucet (access checks)
+    exec.account::get_id dup exec.account::is_non_fungible_faucet assert
+    # => [non_fungible_faucet_id, ASSET]
+
+    # assert that the faucet specified in the asset matches the faucet the transaction is being
+    # executed against
+    dup.3 assert_eq
+    # => [ASSET]
+
+    # validate non-fungible asset by asserting the most significant bit of the most significant
+    # element (ASSET[3]) is 0
+    dup u32split swap drop u32checked_shr.31 not assert dupw
+    # => [ASSET, ASSET]
```

Comment:
Should we just execute `asset::validate_non_fungible_asset` here? I know it is a bit more expensive, but this would be more future proof in case asset validation logic changes in the future.